WIP: ethless vochain 0.1

jordipainan opened the pull request.

jordipainan committed:
vochain: add support for MintTokensTx

Added support for MintTokensTx.
Change Treasurer from hex address to
protobuf Treasurer model and adapt
the existing code accordingly.

jordipainan committed:
vochain: ethless add support for add/del account delegates

jordipainan committed:
vochain: add/del account delegates

Increase nonce on add/del success.
Adds unit testing.

randomshinichi committed:
VPROT-125 TransactionCosts{} represents the tx cost settings in the genesis file (#398)

add transaction costs to the genesis

jordipainan committed:
address minor comments:
- simplify setTreasurer
- delete unrequired boolean from SetAccountDelegateTxCheck()

jordipainan committed:
vochain: add support for sendTokensTx

randomshinichi committed:
State.TxCost()/SetTxCost(); TransactionCosts{} unittests guarantee translation to the keys that State uses

randomshinichi committed:
TransactionCostsFieldFromStateKey() is not needed

rxAlphabetical is unused because TransactionCostsFieldFromStateKey() is removed

randomshinichi committed:
State.SetTxCost(), TxCost() and unittest

randomshinichi committed:
Uint64ToBytes() no longer needed

randomshinichi committed:
cleanup double import

randomshinichi committed:
BaseApplication.InitChain() initializes State.TxCosts from GenesisAppState

jordipainan committed:
vochain: state TxCost Set/Get and Genesis State setting

randomshinichi committed:
golang cleanup

randomshinichi committed:
TestTransactionCostsAsMap() ensure that fields are translated to the correct keys in the Map

randomshinichi committed:
State.SetTxCost(): no need for check, because it will prevent one from setting the cost in the first place

jordipainan committed:
vochain: add support for collectFaucetTx

p4u commented on vochain/account.go:
```diff
@@ -32,15 +34,18 @@ func (a *Account) Unmarshal(data []byte) error {
 
 // Transfer moves amount from the origin Account to the dest Account.
-func (a *Account) Transfer(dest *Account, amount uint64, nonce uint32) error {
+func (a *Account) Transfer(dest *Account, amount uint64, nonce uint32, faucet bool) error {
```
Umm, I don't like this approach (personally). I'd prefer to let Transfer() as a method for transfering balance among accounts and create a new method under `Account` named something like `CollectFaucet()`.

The reason is that Transfer is a very important method that will be used by several parts of the code. Not all developers will understand what the "faucet" boolean means and its easy to make errors (and important ones).

p4u commented on vochain/account.go:
```diff
@@ -89,13 +93,10 @@ func (a *Account) DelDelegate(addr common.Address) error {
 // If provided nonce does not match origin address nonce+1, ErrAccountNonceInvalid is returned.
-// If isQuery is set to true, this method will only check against the current state (no changes will be stored)
-func (v *State) TransferBalance(from, to common.Address, amount uint64, nonce uint32, isQuery bool) error {
+func (v *State) TransferBalance(from, to common.Address, amount uint64, nonce uint64, faucet bool) error {
```
Same as above. I don't like using the same Method for collect faucet funds and send tokens between accounts. I think in this case its better to make things very explicit.

p4u commented on vochain/account.go:
```diff
@@ -392,6 +399,98 @@ func (v *State) setDelegate(accountAddr, delegateAddr common.Address, txType mod
 	}
 }
 
+func SendTokensTxCheck(vtx *models.Tx, txBytes, signature []byte, state *State) (common.Address, common.Address, uint64, uint32, error) {
```
I know returning many individual parameters is a quite common behavior here. However I think we should slowly change it. So IMO it would be better if this method returns a struct+error instead of 5 unamed parameters.

jordipainan committed:
vochain: add tx cost in txs

jordipainan committed:
refactor testing & vochain-ethless:

Refactor all vocdoni-node unit and some
integration testing with the latest changes
added by the vochain-ethless approach

Changes on SendTokensTx and CollectFaucetTx
for better readability and flexibility for
future work

Some minor changes required by the refactor

jordipainan committed:
testsuite: adapt tests to support ethless vochain

jordipainan committed:
add more integration tests

jordipainan committed:
minor fixes and readability improvements:
add missing checks on some getAccount operations
fix ipfssync double import
fix unused scrutinizer test function return value
add accountNotFound error
add wrapper struct for setAccountInfoTxCheck return values
